src/helpers/buffers/unordered_receiver.rs

```diff
@@ -203,30 +215,78 @@ where
 
     /// Poll for the next record.  This should only be invoked when
     /// the future for the next message is polled.
+    ///
+    /// ## Errors
+    /// If buffer capacity is not enough to read `M` and the underlying stream does not have
+    /// more data. This may lead to different behavior depending on the order of issued reads. See
+    /// [`read_order`] test for an example.
+    ///
+    /// [`read_order`]: test::read_order
     fn poll_next<M: Message>(&mut self, cx: &mut Context<'_>) -> Poll<Result<M, Error>> {
-        if let Some(m) = self.spare.read() {
+        // If spare has enough data for us, poll it first
+        // otherwise, poll the underlying stream until it returns pending or it provides enough
+        // data to return a value.
+        let message = self.spare.read();
+        let next = if let Some(m) = message {
+            // this check exists to make sure the inner stream is eventually moved to
```

Comment:
```suggestion
            // This check exists to make sure the inner stream is eventually moved to
```

I don't know if this comment is still accurate.